### DISCUSSION TITLE
Delegated permissions using username/password

### DISCUSSION THREAD

**Discussion by Lewin8687** [pnp/pnpcore#686] [Q&A]
Hi 
I am facing a problem using delegated permissions, but got unauthorized exception.
![image](https://user-images.githubusercontent.com/9344670/145929784-8412c00f-d2f7-4859-a4a0-fde69a81e30b.png)

Stack trace:
   at PnP.Core.Services.BatchClient.<ExecuteSharePointRestInteractiveAsync>d__50.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at PnP.Core.Services.BatchClient.<ExecuteBatch>d__36.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at PnP.Core.Model.BaseDataModel`1.<RequestAsync>d__93.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at PnP.Core.Services.PnPContextFactory.<InitializeContextAsync>d__40.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at PnP.Core.Services.PnPContextFactory.<CreateAsync>d__35.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()

**Here is my setup:**

- Azure app granted with these permissions
![image](https://user-images.githubusercontent.com/9344670/145929335-494437f0-6394-47de-9d8c-d9d87ec26270.png)

- Added user to the enterprise app
- Filled in correct client id/tenant id/username/password

Any advice? Thanks:)

**Comment by jansenbe** [comment]
@Lewin8687 : delegated permissions using username+pwd definitely work. Common reason for failure is either something wrong in your configuration file or some issue in the code that adds PnP Core SDK into your project (the `AddPnPCore` and `AddPnPCoreAuthentication` methods). See https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-cosmosdb-v2-input?tabs=csharp for documentation. We also have one of our samples using the username + password: see https://pnp.github.io/pnpcore/demos/Demo.RPi/README.html

**Reply by Lewin8687** [reply]
@jansenbe  Thanks for the reply!

Unfortunately, I haven't made it work so far.
I was following the steps here: https://pnp.github.io/pnpcore/using-the-sdk/configuring%20authentication.html#delegated-permissions-acting-in-the-name-of-the-user

`AddPnPCoreAuthentication ` is added

```services.AddPnPCoreAuthentication(
                options => {
                    // Configure an Authentication Provider relying on Windows Credential Manager
                    options.Credentials.Configurations.Add("UsernamePassword",
                        new PnPCoreAuthenticationCredentialConfigurationOptions
                        {
                            ClientId = "<filled with correct value>",
                            TenantId = "<filled with correct value>",
                            UsernamePassword = new PnPCoreAuthenticationUsernamePasswordOptions 
                            {
                                Username = "<filled with correct value>",
                                Password = "<filled with correct value>"
                            }
                        });
                    // Configure the default authentication provider
                    options.Credentials.DefaultConfiguration = "UsernamePassword";
                    // Map the site defined in AddPnPCore with the 
                    // Authentication Provider configured in this action
                    options.Sites.Add("SiteToWorkWith",
                        new PnPCoreAuthenticationSiteOptions
                        {
                            AuthenticationProviderName = "UsernamePassword"
                        });
                });```

But I assume I can use either `AddPnPCoreAuthentication` to add authentication method when application starts or `UsernamePasswordAuthenticationProvider` when creating the pnp context right? They both will take azure app delegated permissions, correct?

BTW, was the first link you attached is a mistake, its talking about cosmo db.

**Reply by Lewin8687** [reply]
@jansenbe  Ok i just read something. Looks like in order to access a SharePoint Online site, delegated permissions require not just the app to have **allsites.fullcontrol** permissions, but also the user to have access to the site, its actually taking the intersection of the permissions between the user and the app.

So its impossible to use delegated permissions to allow a user who doesn't have access to a site to access the site.
Please let me know if my understanding is correct, thanks!

**Reply by jansenbe** [reply]
@Lewin8687 : that's correct. Using delegated permissions both the user and the app need access. You can't given the app access to a site the user cannot access. For that use case you'd need Application permissions

**Reply by Lewin8687** [reply]
@jansenbe Thanks for the explaination! I have another question for delegated permissions.

I did some tests to see what exceptions we get for different scenarios. My case is to get the list of classification models.

It all worked until I removed/revoked application permissions **allSites.FullControl**.

Before I added and granted **allSites.FullControl** permission, I got exceptions all the time which is good. 
After I removed or revoked **allSites.FullControl** permission, I can still get the context successfully, there is no exception but NO classification models fetched. Is this expected? I was expecting to get exceptions again.

Thanks in advance!

**Reply by jansenbe** [reply]
@Lewin8687 : might be that the underlying MSAL library is still caching the token, a token lives for 1h. Do you still see the same behavior when you restart your application?

**Reply by Lewin8687** [reply]
@jansenbe  That shouldn't be the case as I restarted the app and I waited for more than 1 hour, it's still happening. Is it possible that once delegated permissions are assigned, even the permissions are removed from the app, the app is still authenticated but not authorized?

Cuz of the experiment I have done, I am able to connect to the site but return 0 items.

**Reply by jansenbe** [reply]
That's kind of weird @Lewin8687 : can you verify the scopes in the access token (e.g. paste the token in https://jwt.io/) and see if the scopes you've removed are gone?

**Reply by Lewin8687** [reply]
@jansenbe  Yes, scopes are correct, sharepoint allsites is removed.  And app Id is set correctly.
![image](https://user-images.githubusercontent.com/9344670/146383666-37a83ba8-9569-4d01-95cf-e3735fd8e7ab.png)

**Reply by jansenbe** [reply]
Might be some internal Syntex API behavior then? If you create a new app do you then again get the exception? Your business logic should be able to handle both scenarios then to be at the safe side.

**Reply by Lewin8687** [reply]
Yes, I guess so. 

I did another test with a new app. I got exceptions when no permissions were added, but after I added and removed permissions, no exceptions anymore but cannot get any item.

It's just a bit tricky to tell whether there is no item or no permissions to get them (I was testing getting classification models in this case).

**Reply by Lewin8687** [reply]
@jansenbe  Sorry, do we have any updates? Thanks!

**Reply by jansenbe** [reply]
@Lewin8687 : no updates from my side here. Guess this is by design
